sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSampledSpansProcessor.java

```diff
@@ -0,0 +1,292 @@
+    private final Object monitor = new Object();
+
+    @GuardedBy("monitor")
+    private final List<ReadableSpan> spansList;
+
+    private Worker(
+        SpanExporter spanExporter,
+        long scheduleDelayMillis,
+        int maxQueueSize,
+        int maxExportBatchSize) {
+      this.spanExporter = spanExporter;
+      this.scheduleDelayMillis = scheduleDelayMillis;
+      this.maxQueueSize = maxQueueSize;
+      this.halfMaxQueueSize = maxQueueSize >> 1;
+      this.maxExportBatchSize = maxExportBatchSize;
+      this.spansList = new ArrayList<>(maxQueueSize);
+    }
+
+    private void addSpan(ReadableSpan span) {
+      synchronized (monitor) {
+        if (spansList.size() == maxQueueSize) {
+          // TODO: Record a counter for dropped spans.
+          return;
+        }
+        // TODO: Record a gauge for referenced spans.
+        spansList.add(span);
+        // Notify the worker thread that at half of the queue is available. It will take
+        // time anyway for the thread to wake up.
+        if (spansList.size() >= halfMaxQueueSize) {
+          monitor.notifyAll();
+        }
+      }
+    }
+
+    @Override
+    public void run() {
+      while (!Thread.currentThread().isInterrupted()) {
+        // Copy all the batched spans in a separate list to release the monitor lock asap to
+        // avoid blocking the producer thread.
+        ArrayList<ReadableSpan> spansCopy;
+        synchronized (monitor) {
+          // If still maxExportBatchSize elements in the queue better to execute an extra
+          if (spansList.size() < maxExportBatchSize) {
+            do {
+              // In the case of a spurious wakeup we export only if we have at least one span in
+              // the batch. It is acceptable because batching is a best effort mechanism here.
+              try {
+                monitor.wait(scheduleDelayMillis);
+              } catch (InterruptedException ie) {
+                // Preserve the interruption status as per guidance and stop doing any work.
+                Thread.currentThread().interrupt();
+                return;
+              }
+            } while (spansList.isEmpty());
+          }
+          spansCopy = new ArrayList<>(spansList);
+          spansList.clear();
+        }
+        // Execute the batch export outside the synchronized to not block all producers.
+        exportBatches(spansCopy);
+      }
+    }
+
+    private void flush() {
+      ArrayList<ReadableSpan> spansCopy;
+      synchronized (monitor) {
+        spansCopy = new ArrayList<>(spansList);
+        spansList.clear();
+      }
+      // Execute the batch export outside the synchronized to not block all producers.
+      exportBatches(spansCopy);
+    }
+
+    private void exportBatches(ArrayList<ReadableSpan> spanList) {
+      ArrayList<Span> spansProtoList = new ArrayList<>(maxExportBatchSize);
+      // TODO: Record a counter for pushed spans.
+      for (int i = 0; i < spanList.size(); i++) {
```

Comment:
Suggestion: consider `Lists.partition(spanList, maxExportBatchSize)` instead of checking whether `spansProtoList` is full every iteration.

Reply:
https://github.com/census-instrumentation/opencensus-java/pull/1882#discussion_r281418009

Reply:
Changed the code to avoid the check all the time.